cl/alerts/management/commands/cl_send_scheduled_alerts.py

```diff
@@ -0,0 +1,123 @@
+# !/usr/bin/python
+# -*- coding: utf-8 -*-
+import datetime
+
+from django.db.models import Prefetch
+from django.http import QueryDict
+from django.utils.timezone import now
+
+from cl.alerts.models import Alert, ParentAlert, UserRateAlert
+from cl.alerts.send_alerts import (
+    merge_highlights_into_result,
+    send_search_alert_and_webhooks,
+)
+from cl.alerts.utils import InvalidDateError
+from cl.lib.command_utils import VerboseCommand, logger
+from cl.search.constants import ALERTS_HL_TAG
+from cl.search.models import SEARCH_TYPES
+from cl.stats.utils import tally_stat
+
+
+def json_date_parser(dct):
+    for key, value in dct.items():
+        if isinstance(value, str):
+            try:
+                dct[key] = datetime.datetime.fromisoformat(value)
+            except ValueError:
+                pass
+    return dct
+
+
+def query_and_send_alerts_by_rate(rate: str) -> None:
+    """Query and send alerts per user.
+
+    :param rate: The alert rate to send Alerts.
+    :return: None
+    """
+
+    alerts_sent_count = 0
+    now_time = now()
+    alerts_to_update = []
+    rate_users = (
+        UserRateAlert.objects.filter(rate=rate)
+        .select_related("user")
+        .prefetch_related(
+            Prefetch(
+                "parent_alerts",
+                queryset=ParentAlert.objects.select_related(
+                    "alert"
+                ).prefetch_related("scheduled_alerts"),
+            )
+        )
+    )
+
+    for rate_user in rate_users.iterator():
+        hits = []
+        for parent_alert in rate_user.parent_alerts.all():
+            results = parent_alert.scheduled_alerts.all()
+            qd = QueryDict(parent_alert.alert.query.encode(), mutable=True)
+            search_type = qd.get("type", SEARCH_TYPES.OPINION)
+            documents = []
+
+            for result in results:
+                document_content = json_date_parser(result.document_content)
+                if result.highlighted_fields:
+                    merge_highlights_into_result(
+                        result.highlighted_fields,
+                        document_content,
+                        ALERTS_HL_TAG,
+                    )
+                documents.append(document_content)
+            alerts_to_update.append(parent_alert.alert)
+            hits.append(
+                (
+                    parent_alert.alert,
+                    search_type,
+                    documents,
+                    len(documents),
+                )
+            )
+        send_search_alert_and_webhooks(rate_user.user, hits)
+        if rate_user.parent_alerts.exists():
+            alerts_sent_count += 1
+
+    # Update Alert's date_last_hit in bulk.
+    Alert.objects.filter(
+        id__in=[alert.id for alert in alerts_to_update]
+    ).update(date_last_hit=now_time)
+
+    # Remove stored alerts sent, all the related objects will be deleted
+    # in cascade.
+    rate_users.delete()
```

Comment:
This command was also refactored. I realized that it's better to only change the status of the `ScheduledAlertHit` instances to `Sent` and only removed them after some time (I set it to 90 days). So in case of debugging we'll have the object in DB.